src/bvals/cc/boundary_communication.cpp

```diff
@@ -0,0 +1,295 @@
+  if (cache.buf_vec.size() == 0)
+    InitializeBufferCache<bound_type>(md, &(pmesh->boundary_comm_map), &cache, ReceiveKey,
+                                      false);
+
+  std::for_each(std::begin(cache.buf_vec), std::end(cache.buf_vec),
+                [](auto pbuf) { pbuf->TryStartReceive(); });
+
+  Kokkos::Profiling::popRegion(); // Task_StartReceiveBoundBufs
+  return TaskStatus::complete;
+}
+
+template TaskStatus
+StartReceiveBoundBufs<BoundaryType::any>(std::shared_ptr<MeshData<Real>> &);
+template TaskStatus
+StartReceiveBoundBufs<BoundaryType::local>(std::shared_ptr<MeshData<Real>> &);
+template TaskStatus
+StartReceiveBoundBufs<BoundaryType::nonlocal>(std::shared_ptr<MeshData<Real>> &);
+
+template <BoundaryType bound_type>
+TaskStatus ReceiveBoundBufs(std::shared_ptr<MeshData<Real>> &md) {
+  Kokkos::Profiling::pushRegion("Task_ReceiveBoundBufs");
+
+  Mesh *pmesh = md->GetMeshPointer();
+  auto &cache = md->GetBvarsCache().GetSubCache(bound_type, false);
+  if (cache.buf_vec.size() == 0)
+    InitializeBufferCache<bound_type>(md, &(pmesh->boundary_comm_map), &cache, ReceiveKey,
+                                      false);
+
+  bool all_received = true;
+  std::for_each(
+      std::begin(cache.buf_vec), std::end(cache.buf_vec),
+      [&all_received](auto pbuf) { all_received = pbuf->TryReceive() && all_received; });
+
+  int ibound = 0;
+  if (Globals::sparse_config.enabled) {
+    ForEachBoundary<bound_type>(
+        md, [&](sp_mb_t pmb, sp_mbd_t rc, nb_t &nb, const sp_cv_t v) {
+          const std::size_t ibuf = cache.idx_vec[ibound];
+          auto &buf = *cache.buf_vec[ibuf];
+
+          // Allocate variable if it is receiving actual data in any boundary
+          // (the state could also be BufferState::received_null, which corresponds to no
+          // data)
+          if (buf.GetState() == BufferState::received && !v->IsAllocated()) {
+            pmb->AllocateSparse(v->label());
+            // TODO(lfroberts): Need to flag this so that the array gets filled with
+            //                  something sensible, currently just defaulted to zero.
+          }
+          ++ibound;
+        });
+  }
+  Kokkos::Profiling::popRegion(); // Task_ReceiveBoundBufs
+  if (all_received) return TaskStatus::complete;
+  return TaskStatus::incomplete;
+}
+
+template TaskStatus
+ReceiveBoundBufs<BoundaryType::any>(std::shared_ptr<MeshData<Real>> &);
+template TaskStatus
+ReceiveBoundBufs<BoundaryType::local>(std::shared_ptr<MeshData<Real>> &);
+template TaskStatus
+ReceiveBoundBufs<BoundaryType::nonlocal>(std::shared_ptr<MeshData<Real>> &);
+
+template <BoundaryType bound_type>
+TaskStatus SetBounds(std::shared_ptr<MeshData<Real>> &md) {
+  Kokkos::Profiling::pushRegion("Task_SetInternalBoundaries");
+
+  Mesh *pmesh = md->GetMeshPointer();
+  auto &cache = md->GetBvarsCache().GetSubCache(bound_type, false);
+
+  auto [rebuild, nbound] = CheckReceiveBufferCacheForRebuild<bound_type, false>(md);
```

Comment:
Are tuple comprehensions a C++17 feature? Because that is really nice.

Reply:
Yeah, although they are not actually necessary, structured bindings are one of the C++17 features I was waiting on most.